[B+C] Add InventoryMoveItemEvent. Adds BUKKIT-3765

TriPhoenix opened the pull request:
**The Issue:**
Currently there is no event to monitor or cancel hoppers moving items between inventories. 

**Justification for this PR:**
This PR implements an InventoryMoveEvent, that can be used to track automatic movements between inventories. This applies to hoppers but may as well be used for droppers when they have a fully functional Inventory interface.

**PR Breakdown:**
This PR simply creates another event type. Source and destination inventories are available as well as the ItemStack being moved and the Block or Entity (as this can be caused by both, blocks like hoppers but as well entities like hopper minecarts) which induced the transfer. 

**Testing Results and Materials:**
A simple test plugin can be found at https://gist.github.com/TriPhoenix/5207263 It simply catches and displayes the events and cancels all events that try to move dirt.

**Relevant PR(s):**
CB-1074 - https://github.com/Bukkit/CraftBukkit/pull/1074 - accompanying implementation for hoppers and hopper minecarts

**JIRA Ticket:**
BUKKIT-3765 - https://bukkit.atlassian.net/browse/BUKKIT-3765

riking commented on src/main/java/org/bukkit/event/inventory/InventoryMoveItemEvent.java:
```diff
@@ -0,0 +1,102 @@
+
+    /**
+     * Gets the Entity that initiated the transfer
+     * 
```
These pesky javadocs, always adding extra whitespace :(

Remember to use `git diff --check` before committing, it should highlight these!

natemort commented on src/main/java/org/bukkit/event/inventory/InventoryMoveItemEvent.java:
```diff
@@ -0,0 +1,102 @@
+    private final Inventory destinationInventory;
+    private final ItemStack itemStack;
+    private final Block initiatorBlock;
+    private final Entity initiatorEntity;
```
You can get the InventoryHolder of an inventory to get the Block or Entity that holds an inventory.  Is there a reason to explicitly keep a reference to these?

TriPhoenix replied:
Agreed, these can be merged into a combined InventoryHolder (which I will do shortly). The basic intention for keeping an initiator though is to determine, which one actually started the transfer as this can be source or destination. While currently one of the two has to be a hopper, this would already be ambigous in events where one hopper fills another hopper (which might be relevant for example in region-based checks).

natemort commented on src/main/java/org/bukkit/event/inventory/InventoryMoveItemEvent.java:
```diff
@@ -0,0 +1,102 @@
+import org.bukkit.inventory.ItemStack;
+
+/**
+ * Called when some entity or block (e.g. hopper) tries to move items from one inventory to another
```
Does this event include items being picked up from a hopper?

TriPhoenix replied:
No, those are handled by the [InventoryPickupItemEvent](http://jd.bukkit.org/dev/apidocs/org/bukkit/event/inventory/InventoryPickupItemEvent.html) already in master. This only handles direct movements between inventories.

natemort replied:
Also, what happens when this event is cancelled?  What's the behavior of the ItemStack and the Inventories?

natemort commented on src/main/java/org/bukkit/event/inventory/InventoryMoveItemEvent.java:
```diff
@@ -0,0 +1,102 @@
+    /**
+     * Gets the Inventory that the ItemStack is being picked from
+     * 
+     * @return Inventory, may be null if the source is no InventoryHolder
```
"not an InventoryHolder" sounds better to me, rather than "is no InventoryHolder".

TriPhoenix commented:
Oh dear, there's always some whitespace yet to miss. I hopefully addressed all relevant issues, feel free to comment further if the documentation is unclear in some points.

One point, that I'd like to explain, is the inclusion of a copied ItemStack rather than a modifieable ItemStack. The current logic in the NMS part would require some rather large code modifications to support this which is (as far as I know) against the minimal-diff policy. Thus I chose the minimal-diff route and skipped supporting a changeable ItemStack.

Otherwise I hope I'm correct to amend this PR with additional commits instead of squashing it all into one :)

amaranth commented:
The source inventory is also the initiator and you should be able to get all relevant information from that, no need for an extra InventoryHolder. Also I'd like to see this include the to/from slots for logging. I know that'll mean a larger diff but I think it should be possible and would be very helpful. If putting the event in a place to get that info also means we can modify the ItemStack then I'd call that a win but that isn't required.

TriPhoenix commented:
Replaced the InventoryHolder with a boolean which is used to determine the correct initiator. Form/to logging and/or ItemStack changing still pending, this will require more changes on the CB side, which probably will wait for the 1.5.1 diff (as there seems to be some NMS method renaming)

TriPhoenix commented:
Normally I'd agree but in this case this is a constructor only referenced in another pull request on the CB side of things right now. Deprecating the other constructors would only clutter the API unnecessarily but would not help compatibility of the API as this is not yet in Bukkit.

natemort commented on src/main/java/org/bukkit/event/inventory/InventoryMoveItemEvent.java:
```diff
@@ -0,0 +1,92 @@
+ * If this event is not cancelled, the initiator will try to move an item (or items) equivalent to the ItemStack
+ * from the source inventory to the destination inventory. If the destination inventory is full, the move will
+ * not take place and the initiator may continue to find another suitable item (which will result in another item).
+ */
```
We limit javadocs lines at 78 characters per line.  If you could fix that here and throughout the rest of the file, that'd be great.

TriPhoenix replied:
Sure thing, might be missing from the contributing.md though, it currently only mentions this limit for commit comments :)

TriPhoenix committed:
Add InventoryMoveItemEvent. Adds BUKKIT-3765

This event is being called whenever a block or entity (e.g. hopper) tries to
move an item from one inventory to another inventory (one inventory may be
the hopper itself).

TriPhoenix commented:
Updated to reflect the possibility of changing the ItemStack and the associated consequences. I hope this clears all corner cases, if not, feel free to inquire :)

riking commented:
(Deleted stupid comment)

natemort commented:
Pulled, thanks.
